packages/loader/container-loader/src/container.ts

```diff
@@ -1682,36 +1680,25 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
     }
 
     private processRemoteMessage(message: ISequencedDocumentMessage): IProcessMessageResult {
-        // Check and report if we're getting messages from a clientId that we previously
-        // flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
-        if (message.clientId != null) {
-            let errorMsg: string | undefined;
-            const client: ILocalSequencedClient | undefined =
-                this.getQuorum().getMember(message.clientId);
-            if (client === undefined && message.type !== MessageType.ClientJoin) {
-                // pre-0.58 error message: messageClientIdMissingFromQuorum
-                errorMsg = "Remote message's clientId is missing from the quorum";
-            } else if (client?.shouldHaveLeft === true && message.type !== MessageType.NoOp) {
-                // pre-0.58 error message: messageClientIdShouldHaveLeft
-                errorMsg = "Remote message's clientId already should have left";
-            }
-            if (errorMsg !== undefined) {
-                const error = new DataCorruptionError(
-                    errorMsg,
-                    extractSafePropertiesFromMessage(message));
-                this.close(normalizeError(error));
-            }
-        }
-
         const local = this.clientId === message.clientId;
 
+        // Allow the protocol handler to process the message
+        let result: IProcessMessageResult = { immediateNoOp: false };
+        try {
+            result = this.protocolHandler.processMessageInQuorum(message, local);
+        } catch (error) {
+            assert(error instanceof Error, "Unexpected error type from the protocol handler");
```

Comment:
I think you are doing it backwards. There is no need to normalize error that is known.
I believe you want to remove assert above and do the following:

```ts
const {message, ...props} = extractSafePropertiesFromMessage(message);
this.close(normalizeError(
                new DataCorruptionError(
                    message,
                    props));
```

Reply:
There needs to be a way to get the error message (which can be either 'missing from quorum' or 'should have left') from `processMessage` so that we preserve the error logic. `error` being anything, the assert is needed afaik for `error.message`. Let me know if there's a more elegant way of doing this.

Reply:
assert will result in closing container with this assert tag - I'm not sure if adds any value.
You will have same data in telemetry if you close container directly.
My example above was wrong, I was thinking about something like that:

```ts
const {msg, ...props} = extractLogSafeErrorProperties(error);
this.close(normalizeError(
                new DataCorruptionError(
                    msg ?? "Unknown error",
                    {
                       ...extractSafePropertiesFromMessage(message),
                       ...props,
                    }));
```

But I think the right way to do so is something like that:
```ts
This.close(wrapError(error, (msg) => new DataCorruptionError(msg, extractSafePropertiesFromMessage(message))));
```

Reply:
@markfields  to the rescue

Reply:
The assert was just a type guard so that I can get the message from the error. But `wrapError` is a way better way of going about it. Thanks!